### PR TITLE
Introduce a common GithubApiService

### PR DESCRIPTION
**Pull request by s0undt3ch**
## Contributor Checklist:

* [x] I have updated the unit tests
* [x] I have created a file in the `master/buildbot/newsfragment` directory (and read the `README.txt` in that directory)
* [ ] I have updated the appropriate documentation

### CODE REVIEW

File: master/buildbot/util/githubapiservice.py [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+from twisted.internet import defer
+
+from buildbot.util.httpclientservice import HTTPClientService
+from buildbot.util.logger import Logger
+from buildbot.util.service import SharedService
+
+log = Logger()
+
+
+class GithubApiService(SharedService):
```
Comment (tardyp) [inline master/buildbot/util/githubapiservice.py]: why is it not a subclass of HTTPClientService?

File: master/buildbot/reporters/github.py [hunk 2]
```diff
@@ -54,14 +55,24 @@ def reconfigService(self, token,
         if baseURL.endswith('/'):
             baseURL = baseURL[:-1]
 
-        self._http = yield httpclientservice.HTTPClientService.getService(
-            self.master, baseURL, headers={
-                'Authorization': 'token ' + token,
-                'User-Agent': 'Buildbot'
-            },
-            debug=self.debug, verify=self.verify)
+        self._gh_api_service = yield self.get_github_api_service(baseURL, token)
```
Comment (tardyp) [inline master/buildbot/reporters/github.py]: I think just _github is a better name.

File: master/buildbot/util/githubapiservice.py (hunk 1 above)
Comment (s0undt3ch) [reply master/buildbot/util/githubapiservice.py]: I believe I saw this kind of usage back when I initially wrote this some months ago by looking around at other services.

I'm happy to subclass it if it's also a shared service and gets the job done.

File: master/buildbot/reporters/github.py (hunk 2 above)
Comment (s0undt3ch) [reply master/buildbot/reporters/github.py]: Will update.

**Comment by s0undt3ch** [comment]
Review comments addressed.

**Comment by tardyp** [comment]
Thanks for fixing that. We now need a bit of unit testing :)

**Comment by s0undt3ch** [comment]
Well, what can we test, and please help me with the how. Ie, make sure the token is passed along in the headers if defined?

Testing test access token generation will require an actual github app installed in an ORG and actually send requests to GitHub...

File: master/buildbot/util/githubapiservice.py [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+        if self._access_tokens.get(user_id, None) is not None:
+            expire_date = datetime.datetime.utcnow() + datetime.timedelta(minutes=5)
+            if self._access_tokens[user_id].expires_at < expire_date:
+                self._access_tokens.pop(user_id)
+        if self._access_tokens.get(user_id, None) is None:
+            body = None
+            if user_id:
+                body = {"user_id": user_id}
+
+            client = yield self.get_http_client()
```
Comment (s0undt3ch) [inline master/buildbot/util/githubapiservice.py]: Bah, old code references...
I'll fix it in a bit.

**Comment by tardyp** [comment]
First you can start by fixing the current tests.. Then you can look at the http_client_service tests to see how we do unit tests.

The idea of unit tests is to run the code, and make sure it works as expected, i.e produce the expected http requests, but not really sending the requests

That is very important and allows us to make sure your code works for every version of python we support

**Comment by s0undt3ch** [comment]
The github changes and hook tests will fail. I'm not grasping how my fake gihub service is to be used instead of the real one.

**Comment by s0undt3ch** [comment]
🎉

File: master/buildbot/util/githubapiservice.py [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+                    code=response.code,
+                    error=body
+                )
+        if self._access_tokens.get(user_id, None) is not None:
+            defer.returnValue(self._access_tokens[user_id].token)
+        else:
+            defer.returnValue(None)
+
+    @defer.inlineCallbacks
+    def get(self, endpoint, **kwargs):
```
Comment (tardyp) [inline master/buildbot/util/githubapiservice.py]: You can see the httpclientservice implementation, which avoids to write 5 times the same thing to implement get, post, put , etc

File: master/buildbot/changes/github.py [hunk 5]
```diff
@@ -138,14 +137,32 @@ def reconfigService(self,
         self.category = category if callable(category) else ascii2unicode(
             category)
 
+        if self._github is None:
+            self._github = yield self.get_github_api_service(baseURL)
```
[the remark this answers is not in the document]
Comment (s0undt3ch) [reply master/buildbot/changes/github.py]: Then I'll hit issues with the test suite. I'll see what I can make out of it.

File: master/buildbot/changes/github.py [hunk 6]
```diff
@@ -138,14 +137,32 @@ def reconfigService(self,
...
+
+    @defer.inlineCallbacks
+    def get_github_api_service(self, base_url):
+        if self.debug:
```
[the remark this answers is not in the document]
Comment (s0undt3ch) [reply master/buildbot/changes/github.py]: Ideally, no params would be passed, they're class attributes ... I just cut the code changes short.
I'll fix it.

File: master/buildbot/reporters/github.py [hunk 7]
```diff
@@ -49,19 +49,27 @@ def reconfigService(self, token,
...
-        self._http = yield httpclientservice.HTTPClientService.getService(
-            self.master, baseURL, headers={
-                'Authorization': 'token ' + token,
-                'User-Agent': 'Buildbot'
-            },
-            debug=self.debug, verify=self.verify)
+        self._github = yield self.get_github_api_service(baseURL, token)
         self.verbose = verbose
 
+    @defer.inlineCallbacks
```
[the remark this answers is not in the document]
Comment (s0undt3ch) [reply master/buildbot/reporters/github.py]: Ok

File: master/buildbot/reporters/github.py [hunk 8]
```diff
@@ -49,19 +49,27 @@ def reconfigService(self, token,
...
+    def get_github_api_service(self, base_url, token):
+        if self.debug:
+            log.info('Instantiating GH API Service')
```
[the remark this answers is not in the document]
Comment (s0undt3ch) [reply master/buildbot/reporters/github.py]: Was probably helpful when I was implementing this, but yeah, nothing much added. I'll remove.

File: master/buildbot/test/fake/githubapiservice.py [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+from twisted.internet import defer
+
+from buildbot.test.fake import httpclientservice as fakehttpclientservice
+from buildbot.util import githubapiservice
+from buildbot.util.logger import Logger
+
+log = Logger()
+
+
+class GithubApiService(fakehttpclientservice.HTTPClientService):
```
[the remark this answers is not in the document]
Comment (s0undt3ch) [reply master/buildbot/test/fake/githubapiservice.py]: Erm....

[the remark this answers is not in the document]
Comment (s0undt3ch) [reply master/buildbot/test/fake/githubapiservice.py]: Erm....

File: master/buildbot/test/unit/test_util_httpclientservice.py [hunk 10]
```diff
@@ -121,6 +121,18 @@ def test_post_auth(self):
                                                                 'Content-Type': 'application/json'
                                                             })
 
+    def test_passed_headers_overwrite_base_headers(self):
```
[the remark this answers is not in the document]
Comment (s0undt3ch) [reply master/buildbot/test/unit/test_util_httpclientservice.py]: Ok.

File: master/buildbot/util/githubapiservice.py (hunk 4 above)
Comment (s0undt3ch) [reply master/buildbot/util/githubapiservice.py]: It was copy/paste from the real service.

**Comment by s0undt3ch** [comment]
Cleaned up and tests passing.

**Comment by s0undt3ch** [comment]
Seems that codecov didn't ran on the last build...

File: master/buildbot/changes/github.py [hunk 11]
```diff
@@ -101,21 +101,17 @@ def reconfigService(self,
...
-
-        http_headers = {'User-Agent': 'Buildbot'}
-        if token is not None:
-            http_headers.update({'Authorization': 'token ' + token})
+        self.debug = debug
+        self.verify = verify
 
-        self._http = yield httpclientservice.HTTPClientService.getService(
-            self.master, baseURL, headers=http_headers)
+        if baseURL is not None and baseURL.endswith('/'):
```
Comment (tardyp) [inline master/buildbot/changes/github.py]: this should be handled automatically by GithubAPIService (as this is a code that we would have to use for all its users)

File: master/buildbot/test/unit/test_reporter_github.py [hunk 12]
```diff
@@ -48,14 +48,13 @@ def setUp(self):
...
-                'Authorization': 'token XXYYZZ',
-                'User-Agent': 'Buildbot'
-            },
+            githubapiservice.GithubApiService.API_ROOT_URL,
             debug=None, verify=None)
+        self._http.updateHeaders({'Authorization': 'token XXYYZZ'})
         sp = self.setService()
         sp.sessionFactory = Mock(return_value=Mock())
         yield sp.setServiceParent(self.master)
+        self.patch(sp._github, '_http', self._http)
```
Comment (tardyp) [inline master/buildbot/test/unit/test_reporter_github.py]: 👍

File: master/buildbot/test/unit/test_util_httpclientservice.py [hunk 13]
```diff
@@ -356,6 +382,15 @@ def content(res):
         dl = [oneReq() for i in range(self.NUM_PARALLEL)]
         yield defer.gatherResults(dl)
 
+    @defer.inlineCallbacks
+    def test_passed_headers_overwrite_base_headers(self):
```
Comment (tardyp) [inline master/buildbot/test/unit/test_util_httpclientservice.py]: still needs the same test for TReq

File: master/buildbot/util/githubapiservice.py [hunk 14]
```diff
@@ -0,0 +1,237 @@
...
+    def _doRequest(self, method, endpoint, **kwargs):
+        client = yield self.get_http_client()
+        request_kwargs = yield self._prepare_request_kwargs(**kwargs)
+        deferred = getattr(client, method)(endpoint, **request_kwargs)
+        deferred.addCallback(self._update_rate_limit_information)
+        response = yield deferred
+        defer.returnValue(response)
+
+    @defer.inlineCallbacks
+    def get(self, endpoint, **kwargs):
```
Comment (tardyp) [inline master/buildbot/util/githubapiservice.py]: You can simplify this code:
```
def get(self, endpoint, **kwargs):
    return self._doRequest('get', endpoint, **kwargs)
```

File: master/buildbot/test/unit/test_util_httpclientservice.py (hunk 13 above)
Comment (s0undt3ch) [reply master/buildbot/test/unit/test_util_httpclientservice.py]: `HTTPClientServiceTestTReqE2E`?

**Pull request closed without merging**